src/module-server/rollup-plugin-container.ts

```diff
@@ -189,16 +225,83 @@ export const createPluginContainer = (plugins: Plugin[]) => {
       return Object.keys(opts).length > 1 ? (opts as { id: string }) : id;
     },
 
-    async transform(code: string, id: string) {
+    async transform(
+      originalCode: string,
+      id: string,
+      inputMap?: DecodedSourceMap | RawSourceMap | string,
+    ) {
+      let code = originalCode;
+      // TODO: if any of the transforms is missing sourcemaps, then there should be no source maps emitted
+      const sourceMaps: (DecodedSourceMap | RawSourceMap)[] = [];
+      if (inputMap)
+        sourceMaps.push(
+          typeof inputMap === 'string' ? JSON.parse(inputMap) : inputMap,
+        );
       for (plugin of plugins) {
         if (!plugin.transform) continue;
-        const result = await plugin.transform.call(ctx as any, code, id);
-        if (!result) continue;
+        try {
+          const result = await plugin.transform.call(ctx as any, code, id);
+          if (!result) continue;
+
+          if (typeof result === 'object') {
+            if (!result.code) continue;
+            code = result.code;
+            if (result.map)
+              sourceMaps.push(
+                typeof result.map === 'string'
+                  ? JSON.parse(result.map)
+                  : result.map,
+              );
+          } else {
+            code = result;
+          }
+        } catch (error) {
+          if (error instanceof ErrorWithLocation) {
+            let line = error.line;
+            let column = error.column || 0;
+            if (sourceMaps.length > 0) {
+              const { SourceMapConsumer } = await import('source-map');
+              const consumer = await new SourceMapConsumer(
+                combineSourceMaps(id, sourceMaps) as any,
+              );
+              const sourceLocation = consumer.originalPositionFor({
+                line,
+                column,
+              });
+              consumer.destroy();
+              if (sourceLocation.line !== null) {
+                line = sourceLocation.line;
+                column = sourceLocation.column || 0;
+                if (sourceLocation.source) {
+                  originalCode = await fs.readFile(
+                    sourceLocation.source,
+                    'utf8',
+                  );
+                }
+              }
+            }
+
+            const frame = createCodeFrame(originalCode, line - 1, column);
+            const message = `[${plugin.name}] ${colors.red(
+              colors.bold(error.message),
+            )}
+
+${colors.red(`${id}:${line}:${column + 1}`)}
```

Comment:
@calebeby Can you help me understand why this line has no indentation? 🤔

Reply:
Since it is in the template literal, if I add indents in the source code, the output will have a bunch of spaces in front of it

Reply:
Oh, it's because it's part of the `const message` template literal, right?

Reply:
Ha! Yep, thanks! 👍

Reply:
Yep!